extensions/sql/processors/QueryDatabaseTable.cpp

```diff
@@ -0,0 +1,453 @@
+    "Query must have no ORDER BY statement.")->supportsExpressionLanguage(true)->build());
+
+static const core::Property s_maxRowsPerFlowFile(
+  core::PropertyBuilder::createProperty("qdbt-max-rows")->isRequired(true)->withDefaultValue<int>(0)->withDescription(
+    "The maximum number of result rows that will be included in a single FlowFile. This will allow you to break up very large result sets into multiple FlowFiles. "
+    "If the value specified is zero, then all rows are returned in a single FlowFile.")->supportsExpressionLanguage(true)->build());
+
+static core::Property s_stateDirectory(
+  core::PropertyBuilder::createProperty("State Directory")->isRequired(false)->withDefaultValue("QDTState")->withDescription("Directory which contains processor state data.")->build());
+
+static const core::Relationship s_success("success", "Successfully created FlowFile from SQL query result set.");
+
+// Probably need to have `FragmentAttributes` like nifi has https://github.com/apache/nifi/blob/master/nifi-commons/nifi-utils/src/main/java/org/apache/nifi/flowfile/attributes/FragmentAttributes.java.
+static const std::string FragmentId = "fragment.identifier";
+static const std::string FragmentIndex = "fragment.index";
+static const std::string FragmentCount = "fragment.count";
+
+// State
+class State {
+ public:
+  State(const std::string& stateDir, const std::string& uuid, std::shared_ptr<logging::Logger> logger)
+    :logger_(logger) {
+    if (!createUUIDDir(stateDir, uuid, filePath_)) {
+      return;
+    }
+
+    filePath_ += "State.txt";
+
+    if (!getStateFromFile()) {
+      return;
+    }
+
+    ok_ = true;
+  }
+
+  State::~State() {
+    if (file_.is_open()) {
+      file_.close();
+    }
+  }
+
+  operator bool() const {
+    return ok_;
+  }
+
+  std::unordered_map<std::string, std::string> mapState() const {
+    return mapState_;
+  }
+
+  bool writeStateToFile(const std::unordered_map<std::string, std::string>& mapState) {
+    file_.seekp(std::ios::beg);
+
+    auto dataSize = 0;
+    for (const auto& el : mapState) {
+      file_ << el.first << '=' << el.second << separator();
+      dataSize += el.first.size() + 1 + el.second.size() + separator().size();
+    }
+
+    // Clear old data with ' '. 
+    for (auto i = dataSize_ - dataSize; i > 0; i--) {
+      file_ << ' ';
+    }
+    dataSize_ = dataSize;
+
+    file_.flush();
+
+    mapState_ = mapState;
+  }
+
+ private:
+   static const std::string& separator() {
+     static const std::string s_separator = "@!qdt!@";
+     return s_separator;
+   }
+
+   bool createUUIDDir(const std::string& stateDir, const std::string& uuid, std::string& dir)
+   {
+     if (stateDir.empty()) {
+       dir.clear();
+       return false;
+     }
+
+     auto dirWithBackslash = stateDir;
+     if (stateDir.back() != '\\') {
+       dirWithBackslash += '\\';
+     }
+
+     dir = dirWithBackslash + "uuid\\" + uuid + "\\";
```

Comment:
It's not going to work on Linux. 
Fileutils has a function to get separator.

Reply:
Fixed.